### ISSUE TITLE
Update rule timeout on request (nft ipv6)

### ISSUE THREAD

**Issue by sblagodatskikh** [miniupnp/miniupnp#663]
Building with iptables package
```
sblagodatskikh@sblagodatskikh-pc:~$ upnpc -6 -m eth1 -A 2001:777:1:0:ac00:0:13:bf4b 2222 2001:888:0:88:4aee:cff:febb:1c25 6666 TCP 1000
upnpc : miniupnpc library test client. (c) 2005-2014 Thomas Bernard
Go to http://miniupnp.free.fr/ or http://miniupnp.tuxfamily.org/
for more information.
List of UPNP devices found on the network :
 desc: http://[2001:888:0:88::1]:41639/rootDesc.xml
 st: urn:schemas-upnp-org:device:InternetGatewayDevice:1

Found valid IGD : http://[2001:888:0:88::1]:41639/
Local LAN ip address : 2001:888:0:88:4aee:cff:febb:1c25
AddPinhole: ([2001:777:1:0:ac00:0:13:bf4b]:2222 -> [2001:888:0:88:4aee:cff:febb:1c25]:6666) / Pinhole ID = 2
sblagodatskikh@sblagodatskikh-pc:~$
---
~ # ip6tables -nvL MINIUPNPD1
Chain MINIUPNPD1 (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 ACCEPT     tcp      pppoe_1 *       2001:777:1:0:ac00:0:13:bf4b  2001:888:0:88:4aee:cff:febb:1c25  tcp spt:2222 dpt:6666
~ # 
---
sblagodatskikh@sblagodatskikh-pc:~$ upnpc -6 -m eth1 -A 2001:777:1:0:ac00:0:13:bf4b 2222 2001:888:0:88:4aee:cff:febb:1c25 6666 TCP 10
upnpc : miniupnpc library test client. (c) 2005-2014 Thomas Bernard
Go to http://miniupnp.free.fr/ or http://miniupnp.tuxfamily.org/
for more information.
List of UPNP devices found on the network :
 desc: http://[2001:888:0:88::1]:41639/rootDesc.xml
 st: urn:schemas-upnp-org:device:InternetGatewayDevice:1

Found valid IGD : http://[2001:888:0:88::1]:41639/
Local LAN ip address : 2001:888:0:88:4aee:cff:febb:1c25
AddPinhole: ([2001:777:1:0:ac00:0:13:bf4b]:2222 -> [2001:888:0:88:4aee:cff:febb:1c25]:6666) / Pinhole ID = 2
sblagodatskikh@sblagodatskikh-pc:~$
```
The message appears in the logs: `~ # Jun 26 16:09:00 [ ERR] miniupnpd[1385]: unrecognized data in lease file`, and the record timeout is updated, its OK
But if you use an assembly with the nftables package, then a new rule will be added, and the timeout of the old rule will not be updated:
```
sblagodatskikh@sblagodatskikh-pc:~$ upnpc -6 -m eth1 -A 2001:777:1:0:ac00:0:13:bf4b 2222 2001:888:0:89:4aee:cff:febb:1c25 6666 TCP 1000
upnpc : miniupnpc library test client. (c) 2005-2014 Thomas Bernard
Go to http://miniupnp.free.fr/ or http://miniupnp.tuxfamily.org/
for more information.
List of UPNP devices found on the network :
 desc: http://[2001:888:0:89::1]:44149/rootDesc.xml
 st: urn:schemas-upnp-org:device:InternetGatewayDevice:1

Found valid IGD : http://[2001:888:0:89::1]:44149/
Local LAN ip address : 2001:888:0:89:4aee:cff:febb:1c25
AddPinhole: ([2001:777:1:0:ac00:0:13:bf4b]:2222 -> [2001:888:0:89:4aee:cff:febb:1c25]:6666) / Pinhole ID = 1
sblagodatskikh@sblagodatskikh-pc:~$ upnpc -6 -m eth1 -A 2001:777:1:0:ac00:0:13:bf4b 2222 2001:888:0:89:4aee:cff:febb:1c25 6666 TCP 15
upnpc : miniupnpc library test client. (c) 2005-2014 Thomas Bernard
Go to http://miniupnp.free.fr/ or http://miniupnp.tuxfamily.org/
for more information.
List of UPNP devices found on the network :
 desc: http://[2001:888:0:89::1]:44149/rootDesc.xml
 st: urn:schemas-upnp-org:device:InternetGatewayDevice:1

Found valid IGD : http://[2001:888:0:89::1]:44149/
Local LAN ip address : 2001:888:0:89:4aee:cff:febb:1c25
AddPinhole: ([2001:777:1:0:ac00:0:13:bf4b]:2222 -> [2001:888:0:89:4aee:cff:febb:1c25]:6666) / Pinhole ID = 2
sblagodatskikh@sblagodatskikh-pc:~$
---
~ # nft list chain ip6 filter MINIUPNPD1
table ip6 filter {
	chain MINIUPNPD1 {
		iif "pppoe_1" th dport 6666 th sport 2222 ip6 daddr 2001:888:0:89:4aee:cff:febb:1c25 ip6 saddr 2001:777:1:0:ac00:0:13:bf4b ip6 nexthdr tcp accept
		iif "pppoe_1" th dport 6666 th sport 2222 ip6 daddr 2001:888:0:89:4aee:cff:febb:1c25 ip6 saddr 2001:777:1:0:ac00:0:13:bf4b ip6 nexthdr tcp accept
	}
}
~ #
```

**Comment by miniupnp** [comment]
@sblagodatskikh could you run `miniupnpd` in verbose mode (with `-vv`  option ?) and tell we which messages you have ?

**Comment by miniupnp** [comment]
@sblagodatskikh could you check with the code from the branch https://github.com/miniupnp/miniupnp/tree/issue-663 ?

thanks

**Comment by miniupnp** [comment]
@sblagodatskikh checkout the branch `issue-663`, change to directory `netfilter_nft` and `$ make && ./testnftpinhole` and tell me what happens ;)

I guess we have some work to do to fix it.

**Comment by miniupnp** [comment]
@svenauhagen ping ;)

**Comment by svenauhagen** [comment]
Hi :).
Have you tried this with an inet table?
The code has been changed to support the inet instead if the seperate ip or ip6 tables.
Inet combines both types into one.
The scripts for nftables also setups the inet table infrastructure specifically.

**Comment by miniupnp** [comment]
@svenauhagen I've failed to test the netfilter_nft code on my linux machine yesterday night.
But from what I've read, I think the remote port is not read properly from the nftable to the `rule_t` structure.
I've added `syslog` to debug the issue.
I think there should be some refactoring to fix this :)

**Comment by svenauhagen** [comment]
@miniupnp alright :)
I am a little short on time at the moment but I think I can find the time to test and refactore the code next week.
I will keep you posted.

**Comment by miniupnp** [comment]
@svenauhagen You don't need to refactor the code right now ;)

But I would like to know what are the syslog() output on a working nftables setup

**Comment by miniupnp** [comment]
what happens when you run `testnftpinhole` from the branch  [issue-663](https://github.com/miniupnp/miniupnp/tree/issue-663)

**Comment by svenauhagen** [comment]
@miniupnp the output is:
add_pinhole(eth0, 2a00::dead:beaf, 1911, fe80::1023:4095, 34952, 6, dummy description, 280989)
rule_set_filter6[1110]: inet filter miniupnpd
  [ meta load iif => reg 1 ]
  [ cmp eq reg 1 0x00000002 ]
  [ payload load 2b @ transport header + 2 => reg 1 ]
  [ cmp eq reg 1 0x00008888 ]
  [ payload load 2b @ transport header + 0 => reg 1 ]
  [ cmp eq reg 1 0x00007707 ]
  [ payload load 16b @ network header + 24 => reg 1 ]
  [ cmp eq reg 1 0x000080fe 0x00000000 0x00000000 0x95402310 ]
  [ payload load 16b @ network header + 8 => reg 1 ]
  [ cmp eq reg 1 0x0000002a 0x00000000 0x00000000 0xafbeadde ]
  [ payload load 1b @ network header + 6 => reg 1 ]
  [ cmp eq reg 1 0x00000006 ]
  [ immediate reg 0 accept ]
testnftpinhole: start_batch[1378]: netlink not connected
testnftpinhole: add_pinhole(): uid=-1
  userdata = { pinholefind_pinhole()
testnftpinhole: refresh_nft_cache[707]: netlink not connected
testnftpinhole: find_pinhole(): uid=-2 desc="" timestamp=0

Anything you are looking for in particular?
This does not look like it is working though,

**Comment by miniupnp** [comment]
@svenauhagen I just added the needed call to init_redirect() ;)

**Comment by svenauhagen** [comment]
@miniupnp some code is missing to use the init_redirect :)

`diff --git a/miniupnpd/netfilter_nft/Makefile b/miniupnpd/netfilter_nft/Makefile
index 8340da7..084d2e4 100644
--- a/miniupnpd/netfilter_nft/Makefile
+++ b/miniupnpd/netfilter_nft/Makefile
@@ -22,7 +22,7 @@ config.h:
 testnftnlrdr:	nftnlrdr.o nftnlrdr_misc.o testnftnlrdr.o $(LIBS)

 testnftpinhole:	nftpinhole.o nftnlrdr_misc.o testnftpinhole.o \
-	../upnputils.o ../linux/getroute.o $(LIBS)
+	../upnputils.o ../linux/getroute.o nftnlrdr.o $(LIBS)

 test_nfct_get:	test_nfct_get.o test_nfct_get.o -lmnl -lnetfilter_conntrack

diff --git a/miniupnpd/netfilter_nft/testnftpinhole.c b/miniupnpd/netfilter_nft/testnftpinhole.c
index 5c96e38..5a134b5 100644
--- a/miniupnpd/netfilter_nft/testnftpinhole.c
+++ b/miniupnpd/netfilter_nft/testnftpinhole.c
@@ -13,6 +13,7 @@
 #include "config.h"
 #include "../miniupnpdtypes.h"
 #include "nftpinhole.h"
+#include "nftnlrdr.h"
 #include "../commonrdr.h"
 #include "../upnputils.h"

@@ -60,7 +61,7 @@ int main(int argc, char * * argv)

 	r = init_redirect();
 	if (r < 0) {
-		syslog(LOG_ERROR, "init_redirect() failed");
+		syslog(LOG_ERR, "init_redirect() failed");
 		return 1;
 	}
`

Here is the output the transport header seems to be a problem:

testnftpinhole: mnl_socket bound, port_id=16422
add_pinhole(eth0, 2a00::dead:beaf, 1911, fe80::1023:4095, 34952, 6, dummy description, 346137)
rule_set_filter6[1110]: inet filter miniupnpd
  [ meta load iif => reg 1 ]
  [ cmp eq reg 1 0x00000002 ]
  [ payload load 2b @ transport header + 2 => reg 1 ]
  [ cmp eq reg 1 0x00008888 ]
  [ payload load 2b @ transport header + 0 => reg 1 ]
  [ cmp eq reg 1 0x00007707 ]
  [ payload load 16b @ network header + 24 => reg 1 ]
  [ cmp eq reg 1 0x000080fe 0x00000000 0x00000000 0x95402310 ]
  [ payload load 16b @ network header + 8 => reg 1 ]
  [ cmp eq reg 1 0x0000002a 0x00000000 0x00000000 0xafbeadde ]
  [ payload load 1b @ network header + 6 => reg 1 ]
  [ cmp eq reg 1 0x00000006 ]
  [ immediate reg 0 accept ]
testnftpinhole: add_pinhole(): uid=1
  userdata = { pinholeget_pinhole_info()
testnftpinhole: table_cb(0xbef21618, 0xbef2262c) filter miniupnpd 2
testnftpinhole: parse_rule_payload: Unsupported payload: (dreg:1, base:TRANSPORT_HEADER, offset:0, len:2)
testnftpinhole: parse_rule_cmp[487]: Unknown cmp (r1type:0, data_len:2, op:0)
end_pinhole_info()
testnftpinhole: get_pinhole_info(1) : :0 => fe80::1023:4095:34952 tcp
testnftpinhole:  desc "pinhole-1 ts-346137: dummy description" ts=346137 packets=0 0
find_pinhole()
testnftpinhole: find_pinhole(): uid=-2 desc="" timestamp=0

**Comment by miniupnp** [comment]
@svenauhagen please try with the updated ` issue-663`  branch (commit 73eb2aaad8b0014a99e669468dc6454c0466890a)

I have rewritten the ` rule_t`  structure

**Comment by svenauhagen** [comment]
@miniupnp this looks very good :)

testnftpinhole: mnl_socket bound, port_id=11870
add_pinhole(eth0, 2a00::dead:beaf, 1911, fe80::1023:4095, 34952, 6, dummy description, 654253)
rule_set_filter6[1105]: inet filter miniupnpd
  [ meta load iif => reg 1 ]
  [ cmp eq reg 1 0x00000002 ]
  [ payload load 2b @ transport header + 2 => reg 1 ]
  [ cmp eq reg 1 0x00008888 ]
  [ payload load 2b @ transport header + 0 => reg 1 ]
  [ cmp eq reg 1 0x00007707 ]
  [ payload load 16b @ network header + 24 => reg 1 ]
  [ cmp eq reg 1 0x000080fe 0x00000000 0x00000000 0x95402310 ]
  [ payload load 16b @ network header + 8 => reg 1 ]
  [ cmp eq reg 1 0x0000002a 0x00000000 0x00000000 0xafbeadde ]
  [ payload load 1b @ network header + 6 => reg 1 ]
  [ cmp eq reg 1 0x00000006 ]
  [ immediate reg 0 accept ]
testnftpinhole: add_pinhole(): uid=1
  userdata = { pinholeget_pinhole_info()
testnftpinhole: table_cb(0xbee73618, 0xbee7462c) filter miniupnpd 2
testnftpinhole: table_cb(0xbee738fc, 0xbee7462c) filter miniupnpd 2
end_pinhole_info()
testnftpinhole: get_pinhole_info(1) : :1911 => fe80::1023:4095:34952 tcp
testnftpinhole:  desc "pinhole-1 ts-654253: dummy description" ts=654253 packets=0 0
find_pinhole()
testnftpinhole: find_pinhole(): uid=1 desc="dummy description" timestamp=654253
get_pinhole_info()
end_pinhole_info()
testnftpinhole: get_pinhole_info(1) : 2a00::dead:beaf:1911 => fe80::1023:4095:34952 tcp
testnftpinhole:  desc "pinhole-1 ts-654253: dummy description" ts=654253 packets=0 0
update_pinhole()
update add_pinhole(eth0, 2a00::dead:beaf, fe80::1023:4095, 1911, 34952, 6, pinhole-1 ts-657793: )
rule_set_filter6[1105]: inet filter miniupnpd
  [ meta load iif => reg 1 ]
  [ cmp eq reg 1 0x00000002 ]
  [ payload load 2b @ transport header + 2 => reg 1 ]
  [ cmp eq reg 1 0x00008888 ]
  [ payload load 2b @ transport header + 0 => reg 1 ]
  [ cmp eq reg 1 0x00007707 ]
  [ payload load 16b @ network header + 24 => reg 1 ]
  [ cmp eq reg 1 0x000080fe 0x00000000 0x00000000 0x95402310 ]
  [ payload load 16b @ network header + 8 => reg 1 ]
  [ cmp eq reg 1 0x0000002a 0x00000000 0x00000000 0xafbeadde ]
  [ payload load 1b @ network header + 6 => reg 1 ]
  [ cmp eq reg 1 0x00000006 ]
  [ immediate reg 0 accept ]
testnftpinhole: update_pinhole(1, ...) returned 0
  userdata = { pinholeget_pinhole_info()
testnftpinhole: table_cb(0xbee73618, 0xbee7462c) filter miniupnpd 2
testnftpinhole: table_cb(0xbee738fc, 0xbee7462c) filter miniupnpd 2
end_pinhole_info()
testnftpinhole: get_pinhole_info(1) : 2a00::dead:beaf:1911 => fe80::1023:4095:34952 tcp
testnftpinhole:  desc "pinhole-1 ts-657793: " ts=657793 packets=0 0
delete_pinhole()
testnftpinhole: delete_pinhole(1) returned 0

**Comment by miniupnp** [comment]
hum it looks like the first call of ` get_pinhole_info`  doesn't fill the remote host...

**Comment by miniupnp** [comment]
@sblagodatskikh @svenauhagen could you test the latest `issue-663` branch ? (commit c5f44ed74537682658060bb070da601194ee765f)

**Comment by miniupnp** [comment]
I also see that `get_pinhole_info` is not filling the `desc` properly. and `update_pinhole` hasn't updated anything :(

**Comment by miniupnp** [comment]
@sblagodatskikh @svenauhagen commit 835a575697fd90eb4fddae42585e6c58c1fca6ff !

**Comment by svenauhagen** [comment]
@miniupnp here is the new output:

testnftpinhole: mnl_socket bound, port_id=10479
add_pinhole(eth0, 2a00::dead:beaf, 1911, fe80::1023:4095, 34952, 6, dummy description, 1036608)
rule_set_filter6[1105]: inet filter miniupnpd
  [ meta load iif => reg 1 ]
  [ cmp eq reg 1 0x00000002 ]
  [ payload load 2b @ transport header + 2 => reg 1 ]
  [ cmp eq reg 1 0x00008888 ]
  [ payload load 2b @ transport header + 0 => reg 1 ]
  [ cmp eq reg 1 0x00007707 ]
  [ payload load 16b @ network header + 24 => reg 1 ]
  [ cmp eq reg 1 0x000080fe 0x00000000 0x00000000 0x95402310 ]
  [ payload load 16b @ network header + 8 => reg 1 ]
  [ cmp eq reg 1 0x0000002a 0x00000000 0x00000000 0xafbeadde ]
  [ payload load 1b @ network header + 6 => reg 1 ]
  [ cmp eq reg 1 0x00000006 ]
  [ immediate reg 0 accept ]
testnftpinhole: add_pinhole(): uid=1
  userdata = { pinholeget_pinhole_info()
testnftpinhole: table_cb(0xbe89f618, 0xbe8a062c) filter miniupnpd 2
testnftpinhole: table_cb(0xbe89f8fc, 0xbe8a062c) filter miniupnpd 2
end_pinhole_info()
testnftpinhole: get_pinhole_info(1) : 2a00::dead:beaf:1911 => fe80::1023:4095:34952 tcp
testnftpinhole:  desc "dummy description" ts=1036608 packets=0 0
find_pinhole()
testnftpinhole: find_pinhole(): uid=1 desc="dummy description" timestamp=1036608
get_pinhole_info()
end_pinhole_info()
testnftpinhole: get_pinhole_info(1) : 2a00::dead:beaf:1911 => fe80::1023:4095:34952 tcp
testnftpinhole:  desc "dummy description" ts=1036608 packets=0 0
update_pinhole()
update add_pinhole(eth0, 2a00::dead:beaf, fe80::1023:4095, 1911, 34952, 6, pinhole-1 ts-1040148: dummy description)
rule_set_filter6[1105]: inet filter miniupnpd
  [ meta load iif => reg 1 ]
  [ cmp eq reg 1 0x00000002 ]
  [ payload load 2b @ transport header + 2 => reg 1 ]
  [ cmp eq reg 1 0x00008888 ]
  [ payload load 2b @ transport header + 0 => reg 1 ]
  [ cmp eq reg 1 0x00007707 ]
  [ payload load 16b @ network header + 24 => reg 1 ]
  [ cmp eq reg 1 0x000080fe 0x00000000 0x00000000 0x95402310 ]
  [ payload load 16b @ network header + 8 => reg 1 ]
  [ cmp eq reg 1 0x0000002a 0x00000000 0x00000000 0xafbeadde ]
  [ payload load 1b @ network header + 6 => reg 1 ]
  [ cmp eq reg 1 0x00000006 ]
  [ immediate reg 0 accept ]
testnftpinhole: update_pinhole(1, ...) returned 0
  userdata = { pinholeget_pinhole_info()
testnftpinhole: table_cb(0xbe89f618, 0xbe8a062c) filter miniupnpd 2
testnftpinhole: table_cb(0xbe89f8fc, 0xbe8a062c) filter miniupnpd 2
end_pinhole_info()
testnftpinhole: get_pinhole_info(1) : 2a00::dead:beaf:1911 => fe80::1023:4095:34952 tcp
testnftpinhole:  desc "dummy description" ts=1040148 packets=0 0
delete_pinhole()
testnftpinhole: delete_pinhole(1) returned 0

**Comment by miniupnp** [comment]
I think we should check that I haven't broken the nat rules code `nftnlrdr.c`

**Comment by miniupnp** [comment]
@svenauhagen could you run `testnftnlrdr` ?

**Comment by svenauhagen** [comment]
@miniupnp I think something broke now

testnftnlrdr: mnl_socket bound, port_id=13228
trying to redirect port 10000 to 192.168.155.220:43333
add redirect rule2((null), (null), 10000, 192.168.155.220, 43333, 6, (null))!
rule_set_dnat[1017]: inet filter prerouting_miniupnpd
  [ payload load 1b @ network header + 9 => reg 1 ]
  [ cmp eq reg 1 0x00000006 ]
  [ payload load 2b @ transport header + 2 => reg 1 ]
  [ cmp eq reg 1 0x00001027 ]
  [ immediate reg 1 0xdc9ba8c0 ]
  [ immediate reg 2 0x000045a9 ]
  [ nat dnat ip addr_min reg 1 addr_max reg 1 proto_min reg 2 proto_max reg 2 ]
testnftnlrdr: send_batch: mnl_cb_run returned -1
testnftnlrdr: nft_send_rule(0xab08b0, 6, 2) send_batch failed -4
addnatrule() failed

**Comment by miniupnp** [comment]
@svenauhagen I think I have not changed anything that could affect `addnatrule()`... could you compare with the master branch ?

btw thanks a lot for your time !

**Comment by svenauhagen** [comment]
:) no I do not think so. I am packed with work right now but I will give it a go during the holidays when I am off work ;)

**Comment by svenauhagen** [comment]
@miniupnp sorry for the delay on this.
I have tested the latest tagged version and it work fine in my nftables setup.

**Comment by miniupnp** [comment]
@svenauhagen I'm sorry, just to make sure could you have a final test before I merge the branch ` issue-663`  ?

I have rebased a maybe some changes were lost or something.

**Comment by svenauhagen** [comment]
@miniupnp yes sorry let me try it today or tomorrow :) I was pulled into other projects the last days.
I will give you my feedback

**Comment by miniupnp** [comment]
branch `issue-663` merged with 177c3ad91552bb718900b53467ba95851cf54739
